README.md
# NO LONGER MAINTAINED

### This project is no longer maintained! If some wants to take over the project and begin merging in PRs, addressing Issues, and updating functionality please reach out!

### Last known compatibility is with Angular 1.5.8
### Thanks you everyone who used and helped create Angular Cron Jobs!


[![Build Status](https://travis-ci.org/angular-cron-jobs/angular-cron-jobs.svg?branch=master)](https://travis-ci.org/angular-cron-jobs/angular-cron-jobs)  [![Coverage Status](https://coveralls.io/repos/github/angular-cron-jobs/angular-cron-jobs/badge.svg?branch=master)](https://coveralls.io/github/angular-cron-jobs/angular-cron-jobs?branch=master)  [![Codacy Badge](https://api.codacy.com/project/badge/Grade/31435876fa31464b8fb495de9b7f49c2)](https://www.codacy.com/app/jc_2/angular-cron-jobs?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=angular-cron-jobs/angular-cron-jobs&amp;utm_campaign=Badge_Grade)  [![npm version](https://badge.fury.io/js/angular-cron-jobs.svg)](https://badge.fury.io/js/angular-cron-jobs)
# angular-cron-jobs
UI Component For Creating Cron Job Syntax To Send To Server

## [Demo](http://angular-cron-jobs.github.io/angular-cron-jobs/#/)

## Installation

Install using bower:

`bower install angular-cron-jobs`

## Use:

Include the component in your application:

    angular.module('myApp', ['angular-cron-jobs']);

Insert the directive where you would like it to appear in your application:

    <cron-selection ng-model="myOutput"></cron-selection>

By setting the ng-model attribute equal to a value in your controller (i.e. `$scope.myOutput` in the example above) you have access to the cron syntax output.  

For example, a job selected to run every month on the 11th at 4:10 AM would output the follow:

    '10 4 11 * *'

as a string.

## Configuration:

The directive takes an optional attribute of `config`

    <cron-selection ng-model="myOutput" config="myConfig"></cron-selection>
    
### Options

This is an object in your controller you can use to remove options from the user.  For example if you would like the user to be able to set Minute, Hour, and Day but not Week, Month, and Year you would create the following object in your controller:

    $scope.myConfig = {
        options: {
            allowWeek : false,
            allowMonth : false,
            allowYear : false
        }
    }

Currently the config object accepts an options property with an object of allowed selections.  These include:

* allowMinute
* allowHour
* allowDay
* allowWeek
* allowMonth
* allowYear

Setting the keys as booleans will turn the selection on and off.

### allowMultiple

You can also set whether or not you want to allow a user to select multiple calues for a cron:

    $scope.myConfig = {
        allowMultiple: true
    }

Setting allowMultiple to either true or false will toggle the ability.

### Quartz by [RajanRastogi](https://github.com/RajanRastogi)

By default, the cron expression generated is Unix Cron compatible. Incase you use the [Quartz Scheduler](https://github.com/quartz-scheduler/quartz), you would want to enable the `quartz` mode in the options.
You can do so by passing that flag as `true` in the config object.

    $scope.myConfig = {
        quartz: true
    }

In case you are using a custom template, please be sure to pass an extra argument `cronStyle` to the `cronDayName` filter. So your code should look something like this:

    value as (value | cronDayName: cronStyle) for value in dayValues

A complete config object may look like the following:

    $scope.myConfig = {
        allowMultiple: true,
        options: {
            allowWeek : false,
            allowMonth : false,
            allowYear : false
        },
        quartz: true
    }

## Custom Templates:

As noted by [TimotheeJeannin](https://github.com/TimotheeJeannin) you can use custom template by setting the template attribute on your cron DOM element:

    <cron-selection template="path/to/my/template.html"></cron-selection>

## Initializing UI with data from server

The directive takes an attribute of `ng-model` used for init and output data.

    <cron-selection ng-model="myOutput" config="myConfig"></cron-selection>

This is a string in your controller of cron syntax that was recieved from your server or any other source:

    $scope.myOutput = "30 2 4 * *"
    
Thew directive will properly build out the UI to reflect this data.

## Setting Cron after directive load

The `ng-model` attribute also works as a reset attribute

    <cron-selection ng-model="myOutput" config="myConfig"></cron-selection>

This is an expression paired with a value in your controller.  Whenever the value changes (or is set for the first time) and passed the `angular.isDefined()` method the cron will reset itself to match that value

    $timeout(function(){
       $scope.myOutput = "0 0 * * *"
    }, 3000);
    
The directive will properly build out the UI to reflect this data.

## Utilize Exposed $scope.myFrequency:

the `frequency` attribute grants you exposure to the `$scope.myFrequency` object inside the directive.  It is two way bound so you can manipulate it from outside the directive as well.

  <cron-selection ng-model="myOutput" config="myConfig" frequency="cronData"></cron-selection>

The type of `{number|Array.<number>}` depends on the following: number if `allowMultiple` is `false` and Array if `allowMultiple` is `true`

Properties you now have access to via `frequency` attribute include:

* **base** `{number}` 1-6 (minute thru year)
* **minuteValue** `{number|Array.<number>}` 0-55 (increments of 5)
* **hourValue** `{number|Array.<number>}` 1-23
* **dayOfMonthValue** `{number|Array.<number>}` 0-6 (Sunday thru Saturday)
* **monthValue** `{number|Array.<number>}` 1-12

### You can get away from using select inputs in your custom temple using the `frequency` attribute.

#### [Demo Without Using Selects](http://angular-cron-jobs.github.io/angular-cron-jobs/#/#noSelectSample)

## Contributors

[@jacobscarter](https://github.com/jacobscarter) - Author

[@wowo](https://github.com/wowo)

[@immertreu](https://github.com/immertreu)

[@TSteele27](https://github.com/TSteele27)

[@DmitryEfimenki](https://github.com/DmitryEfimenko)


## ~Coming Soon:~

~The next big to-do's on my list include:~

~* Support generlized selections such as a one button click for "Every Five Minutes" or "Last Thursday of Every Month"~


CHANGELOG.md
<a name="3.2.1"></a>
# 3.2.1 (2019-09-18)

## Updates

- Hard code dependency versions in bower.json
- Update README.md with end of life message

<a name="3.2.0"></a>
# 3.2.0 (2016-09-19)

## Updates

- Added `quartz` property to config to allow for [Quartz](https://github.com/quartz-scheduler/quartz) style output.
- Address Issue [#24](https://github.com/angular-cron-jobs/angular-cron-jobs/issues/24).

<a name="3.1.1"></a>
# 3.1.1 (2016-09-17)

## Updates

- Revert HTML Tempalte

<a name="3.1.0"></a>
# 3.1.0 (2016-09-17)

## Updates

- Add scripts section to package.json.
- Add ngMultiple directive to reduce template duplication.
- Add baseFrequency value object to reduce dependency on hard-coded numbers.

## Updates

- Revert cleanup because of unexpected breaks.

<a name="3.0.5"></a>
# 3.0.5 (2016-09-17)

## Updates

- Revert cleanup because of unexpected breaks.

<a name="3.0.4"></a>
# 3.0.4 (2016-09-14)

## Updates

- Code cleanup


<a name="3.0.1"></a>
# 3.0.1 (2016-07-18)

## Updates

- Bug fixes in default tempalte.
- Bug fix with allowMultiple reference inside directive.

<a name="3.0.0"></a>
# 3.0.0 (2016-07-18)

## Updates

- Combined `init` and `output` to use Angular's `ngModel`.
- Updated variable names to make more sense.
- Updates to single vs. multiple config, `$scope.frequency` now has numbers for single and array for multiple.

<a name="2.1.1"></a>
# 2.1.1 (2016-06-30)

## Updates

- Correcting build file error.


<a name="2.1.0"></a>
# 2.1.0 (2016-06-30)

## Updates

- Exposed `$scope.myFrequency` by adding two way bound `frequency` attribute to directive.  See README.md for details.


<a name="2.0.0"></a>
# 2.0.0 (2016-06-27)

## Updates

- new major version because tempaltes built to work with 1.4.2 will not work with this version.
- added new config propert of `allowMultiple` which will allow for multiple cron options (e.g '* * 1,3,4 * *').
- update README.md with new multiselect feature and template instructions.
- updated gh-pages demo.


<a name="1.4.2"></a>
# 1.4.2 (2015-04-23)

## Updates

- wrapped each step: CSS wrapper was added to the directive as a whole and each step for easier styling.
- changed filter name: Filter names were polluting name space so the prefix `cron` was added.

src/cronSelection.js
(function(angular){

    "use strict";

    angular.module("angular-cron-jobs", ["templates-angularcronjobs"]);

    angular.module("angular-cron-jobs")
        .directive("cronSelection", ["cronService", "i18nRetriever", "baseFrequency",
            function(cronService, i18n, baseFrequency) {
                return {
                    restrict: "EA",
                    replace: true,
                    transclude: true,
                    require: "ngModel",
                    scope: {
                        ngModel: "=",
                        config: "=",
                        myFrequency: "=?frequency"
                    },
                    templateUrl: function(element, attributes) {
                        return attributes.template || "cronselection.html";
                    },
                    link: function($scope, $el, $attr, $ngModel) {

                        var modelChanged = false;

                        if (typeof $scope.config === "object" && !$scope.config.length) {

                            if (angular.isDefined($scope.config.allowMultiple)) {
                                $scope.allowMultiple = $scope.config.allowMultiple;
                            } else {
                                $scope.allowMultiple = false;
                            }

                            if (angular.isDefined($scope.config.quartz) && $scope.config.quartz) {
                                $scope.cronStyle = "quartz";
                            } else {
                                $scope.cronStyle = "default";
                            }

                            if (angular.isDefined($scope.config.i18n)) {
                                $scope.cronI18n = $scope.config.i18n;
                            } else {
                                $scope.cronI18n = "EN";
                            }

                            i18n.initialize($scope.cronI18n);

                            $scope.templateKeys = i18n.getTemplateKeys();
                        }

                        $scope.baseFrequency = baseFrequency;
                        $scope.frequency = i18n.getFrequencies();

                        if (typeof $scope.config === "object" && !$scope.config.length) {
                            if (typeof $scope.config.options === "object") {
                                var optionsKeyArray = Object.keys($scope.config.options);
                                for (var i in optionsKeyArray) {
                                    var currentKey = optionsKeyArray[i].replace(/^allow/, "");
                                    var originalKey = optionsKeyArray[i];
                                    if (!$scope.config.options[originalKey]) {
                                        for (var b in $scope.frequency) {
                                            if ($scope.frequency[b].label === currentKey) {
                                                $scope.frequency.splice(b, 1);
                                            }
                                        }
                                    }
                                }
                            }
                        }

                        $scope.$watch("ngModel", function (newValue) {
                            if (angular.isDefined(newValue) && newValue) {
                                modelChanged = true;
                                $scope.myFrequency = cronService.fromCron(newValue, $scope.allowMultiple, $scope.cronStyle);
                            } else if (newValue === "") {
                                $scope.myFrequency = undefined;
                            }
                        });

                        $scope.minuteValues = [0, 5, 10, 15, 20, 25, 30, 35, 40, 45, 50, 55];
                        $scope.hourValues = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23];
                        $scope.dayOfMonthValues = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31];
                        $scope.dayValues = [0, 1, 2, 3, 4, 5, 6];
                        $scope.monthValues = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12];

                        if ($scope.cronStyle === "quartz") {
                            $scope.dayValues = [1, 2, 3, 4, 5, 6, 7];
                        }

                        $scope.$watch("myFrequency", function (n, o) {
                            if (n !== undefined) {
                                if (n && n.base && (!o || n.base !== o.base) && !modelChanged) {
                                    setInitialValuesForBase(n);
                                } else if (n && n.base && o && o.base) {
                                    modelChanged = false;
                                }

                                var newVal = cronService.setCron(n, $scope.cronStyle);
                                $ngModel.$setViewValue(newVal);
                            }
                        }, true);

                        function setInitialValuesForBase(freq) {

                            freq.base = parseInt(freq.base);

                            if (freq.base >= baseFrequency.hour) {
                                freq.minuteValues = $scope.minuteValues[0];
                            }

                            if (freq.base >= baseFrequency.day) {
                                freq.hourValues = $scope.hourValues[0];
                            }

                            if (freq.base === baseFrequency.week) {
                                freq.dayValues = $scope.dayValues[0];
                            }

                            if (freq.base >= baseFrequency.month) {
                                freq.dayOfMonthValues = $scope.dayOfMonthValues[0];
                            }

                            if (freq.base === baseFrequency.year) {
                                freq.monthValues = $scope.monthValues[0];
                            }
                        }
                    }
                };
            }]).filter("cronNumeral", ["i18nRetriever", function(i18n) {
        return function(input) {
            if (input !== null) {
                return i18n.getNumeral(input);
            }
            return null;
        };
    }]).filter("cronMonthName", ["i18nRetriever", function(i18n) {
        return function(input) {
            var months = i18n.getMonths();

            if (input !== null && angular.isDefined(months[input])) {
                return months[input];
            } else {
                return null;
            }
        };
    }]).filter("cronDayName", ["i18nRetriever", function(i18n) {
        return function(input, cronType) {

            var days = i18n.getDays();
            var dayDefinition = (cronType === "quartz") ? days[input-1] : days[input];

            if (input !== null && angular.isDefined(dayDefinition)) {
                return dayDefinition;
            } else {
                return null;
            }
        };
    }]).directive("ngMultiple", function() {
        return {
            restrict: "A",
            scope: {
                ngMultiple: "="
            },
            link: function (scope, element) {
                var unwatch = scope.$watch("ngMultiple", function(newValue) {
                    if (newValue) {
                        element.attr("multiple", "multiple");
                    } else {
                        element.removeAttr("multiple");
                    }
                });
            }
        };
    });

}(angular));